### DISCUSSION TITLE
RecordMapper maps Java lang Records using property indexes

### DISCUSSION THREAD

**Discussion by onomesotu** [jOOQ/jOOQ#15568] [Q&A] [answered]
I was using Lombok data annotations for creating a DTO which I map records from a multi-join query into. 

This is the query:
```
  public Optional<UserPrincipalDTO> findByUsername(String email) {
    return dslContext
        .select(
            MEMBER_PROFILE.USERNAME,
            CREDENTIAL.PASSWORD,
            MEMBER_PROFILE.ACTIVE,
            DSL.field(
                    dslContext
                        .select(arrayAgg(AUTHORITIES.AUTHORITY))
                        .from(AUTHORITIES)
                        .where(AUTHORITIES.EMAIL.eq(email)))
                .as("authorities")
        )
        .from(MEMBER)
        .innerJoin(MEMBER_PROFILE)
        .on(MEMBER_PROFILE.MEMBER_ID.eq(MEMBER.ID))
        .innerJoin(CREDENTIAL)
        .on(CREDENTIAL.MEMBER_ID.eq(MEMBER.ID))
        .where(MEMBER.EMAIL.eq(email))
        .fetchOptionalInto(UserPrincipalDTO.class);
  }
```

Below is the Java record:
```
public record UserPrincipalDTO(
    String username, String password, boolean active, List<String> authorities) {}
```

In the above format, it correctly fetches and maps into the `UserPrincipalDTO`, but if I switch the positions of the authorities list and username to this:
```
public record UserPrincipalDTO(
    List<String> authorities, String username, boolean active,  String password) {}
```

It fails because the mappers use the index of the records to map the fields. An excerpt is shown below from DefaultRecordMapper.java L1135 
```
   private final Object[] mapNonnested(R record) {

            // [#10425] Initialise array to constructor parameter type init values
            Object[] converted = Tools.map(parameterTypes, c -> Reflect.initValue(c), Object[]::new);

            for (int i = 0; i < record.size(); i++)
                set(record, i, converted, propertyIndexes[i]);

            return converted;
        }
```
This is very surprising. If this is intentional, can the docs be updated to point out this caveat?

**Comment by lukaseder** [comment, marked as answer]
> This is very surprising.

Well, Java doesn't have named parameters, so there's no way of identifying what you mean by your list of parameters.

Unless you're using a `record`, of course, in case of which the reflection API maintains named `RecordComponent` ordering.

> If this is intentional, can the docs be updated to point out this caveat?

What needs updating?
https://www.jooq.org/javadoc/latest/org.jooq/org/jooq/impl/DefaultRecordMapper.html
